Source/JavaScriptCore/b3/B3ReduceStrength.cpp

```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+            if (m_value->type() == Int32
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 32
+                && shl->child(0) == shr->child(0)) {
+                replaceWithNew<Value>(RotR, m_value->origin(), shr->child(0), shr->child(1));
+                return true;
+            }
+
+            if (m_value->type() == Int64
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 64
+                && shl->child(0) == shr->child(0)) {
+                replaceWithNew<Value>(RotR, m_value->origin(), shr->child(0), shr->child(1));
+                return true;
+            }
+
+            // Otherwise, see if one shift has a Sub operand. We also accept a Sub
+            // that is the single argument of a Trunc node - because this operand is
+            // the shift amount, it semantically acts masked anyway, so the observed
+            // behavior shouldn't change when truncated. This helps when trying to
+            // match this pattern against 64-bit expressions.
+            Value* sub = nullptr;
+            bool isRotL = false;
+            bool isRotR = false;
+            bool needsTrunc = false;
+            if (shl->child(1)->opcode() == Sub) {
+                isRotR = true;
+                sub = shl->child(1);
+            } else if (shl->child(1)->opcode() == Trunc && shl->child(1)->child(0)->opcode() == Sub) {
+                isRotR = true;
+                sub = shl->child(1)->child(0);
+                needsTrunc = true;
+            }
+            if (shr->child(1)->opcode() == Sub) {
+                isRotL = true;
+                sub = shr->child(1);
+            } else if (shr->child(1)->opcode() == Trunc && shr->child(1)->child(0)->opcode() == Sub) {
+                isRotL = true;
+                sub = shr->child(1)->child(0);
+                needsTrunc = true;
+            }
+
+            // Continue only if we have a Sub in exactly one of the two shifts, and
+            // it is subtracting from 32 or 64 respectively based on the type of
+            // this instruction.
+            if (isRotR != isRotL
+                && ((m_value->type() == Int64 && sub->child(0)->isInt(64))
+                    || (m_value->type() == Int32 && sub->child(0)->isInt(32)))) {
+                Value* x = shl->child(0);
+
+                // Check if the x used in each shift is the same.
+                if (x != shr->child(0))
+                    return false;
+
+                // Check if the n used in each shift is the same. As above, since
+                // these expressions are being used as the shift amount, all of this
+                // arithmetic is implicitly mod 32 or mod 64 - so we can safely
+                // ignore intermediate Trunc nodes.
+                Value* n = sub->child(1);
```

Comment:
This `n` is `x2` right? Then, it must be a constant since we have to check the overflow.

Reply:
We also need to check whether constant folding is applied or not by this phase. If it is, then the `sub` case should not be observable here.

Reply:
I'll change this name too, but in this case it's specifically non-constant - or, well, I suppose it could be, but we'd expect the `Sub` to fold away if it was. I think since B3 guarantees wrapping semantics for `Sub`, it's okay if the difference overflows - since overflow here is basically wrapping the shift amount by a large power-of-two modulus, and then the shift itself wraps the shift amount to a smaller power-of-two modulus, I think the overall result can't be observably different.